# Call Center Cleanup

For this assignment, we will be working with call center data. You can start working on the assignment after the first lesson on Exploratory Data Analysis. Make sure to read the whole assignment before starting anything! As you code along in the Jupyter notebook, you are asked to make note of the results of your analysis. Do so by clicking on the results box and adding your notes beside each question.

## Business Issue and Understanding

You are working for a company that has two call centers: the North Call Center and the South Call Center. The company is looking to possibly hire five additional reps to enhance customer experience. Your task is to explore how efficient the current reps are in each branch to determine which branch would benefit from additional assistance.

### How the Call Center Works

Call center representatives are assigned queues. When calls are assigned to a queue, the call is assigned to the next person in line in the queue. After a call is assigned to a representative, the amount of time between assignment and the call starting is divided into busy minutes and not ready minutes. If the call is incoming and a customer is waiting on the phone for a rep, the time is split into three categories: busy minutes, not ready minutes, and incoming call wait time. Once the rep has the customer on the phone, there might be during call wait time, where the call is put on hold while the rep gets an answer for the customer.

### Notes about the Dataset

If you haven't worked in a call center before, these notes might help you throughout your analysis.

* The call purpose is tagged for each call.
* The time of the call is tagged in 1 hour blocks starting at 9:00 AM and ending at 5:00 PM.
* Calls are tagged as incoming or outgoing.
* Reps are assigned to queues. When the reps are working, they take calls in the order of their queue.
* A call that is dropped due to technical error or missed by the center because they have reached maximum capacity is a lost call.
* An abandoned call is when the customer hangs up because they have been waiting for too long.
* Busy Minutes: the amount of time after a call comes in or needs to go out where the assigned rep is not available because they are busy with other customers.
* Not Ready Minutes: the amount of time after a call comes in or needs to go out where the assigned rep is not available because they are not ready (for example, getting water).
* Incoming Wait Time - amount of time after assigned rep is available to take the call customer waits for representative to pick up a call. This is tracked in seconds.
* During Call Wait Time - amount of time during call that customer has to wait for representative

## Getting Started

You have two CSVs at your disposal, `NorthCallCenter.csv` and `SouthCallCenter.csv`. Import the appropriate libraries and create two dataframes, one called `north_df` and one called `south_df`.

In [1]:
# Import the appropriate libraries with aliases
import pandas as pd
# Create two new dataframes
north_df=pd.read_csv("NorthCallCenter.csv")
north_df.info()
north_df.head(20)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 245 entries, 0 to 244
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             245 non-null    int64  
 1   Branch                 245 non-null    object 
 2   Call Purpose           245 non-null    object 
 3   Time Block             245 non-null    object 
 4   Incoming or Outgoing   245 non-null    object 
 5   Queue                  245 non-null    object 
 6   Rep ID                 245 non-null    object 
 7   Sale                   245 non-null    object 
 8   Lost Call              245 non-null    int64  
 9   Abandoned              245 non-null    int64  
 10  Busy Minutes           245 non-null    int64  
 11  Not Ready Minutes      245 non-null    int64  
 12  Incoming Wait Time     82 non-null     float64
 13  During Call Wait Time  245 non-null    int64  
 14  Calls                  245 non-null    int64  
dtypes: flo

Unnamed: 0 Branch     Call Purpose Time Block Incoming or Outgoing Queue  \
0            0  North        Complaint    9:00 AM             Incoming     B   
1            1  North    Sales Support   11:00 AM             Incoming     A   
2            2  North    Sales Support   11:00 AM             Incoming     B   
3            3  North    Sales Support    4:00 PM             Incoming     B   
4            4  North  Product Support   12:00 PM             Incoming     B   
5            5  North  Product Support    2:00 PM             Incoming     B   
6            6  North  Product Support    2:00 PM             Incoming     A   
7            7  North  Product Support    2:00 PM             Incoming     B   
8            8  North  Product Support    3:00 PM             Incoming     B   
9            9  North  Product Support    3:00 PM             Incoming     A   
10          10  North  Product Support    4:00 PM             Incoming     A   
11          11  North  Product Support    4:00 PM             Incoming     A   
12          12  North  Product Support    4:00 PM             Incoming     B   
13          13  North        Complaint    9:00 AM             Incoming     B   
14          14  North        Complaint   12:00 PM             Incoming     B   
15          15  North        Complaint    1:00 PM             Incoming     B   
16          16  North        Complaint    3:00 PM             Incoming     B   
17          17  North        Complaint    2:00 PM             Incoming     A   
18          18  North    Sales Support   10:00 AM             Incoming     B   
19          19  North    Sales Support   10:00 AM             Incoming     B   

    Rep ID  Sale  Lost Call  Abandoned  Busy Minutes  Not Ready Minutes  \
0    Brent    NO          0          1             9                  2   
1      Cam    NO          0          0            11                  1   
2     Todd    NO          0          0             9                  3   
3    Brent    NO          0          0            11                  2   
4    Brent    NO          0          0             8                  2   
5     Duke    NO          0          0            12                  2   
6      Joe    NO          0          0             8                  1   
7    Lilly    NO          0          0            11                  2   
8     Duke    NO          0          0            12                  2   
9      Cam    NO          0          0            11                  3   
10     Joe    NO          0          0            11                  3   
11  Amanda    NO          0          0            10                  3   
12    Todd    NO          0          0            12                  2   
13    Duke    NO          0          0             9                  2   
14   Brent    NO          0          0             8                  2   
15    Duke    NO          0          0            12                  1   
16   Brent    NO          0          0             9                  3   
17    Andy    NO          0          1             9                  3   
18    Duke    NO          0          0            12                  2   
19   Lilly  YES           0          0            11                  2   

    Incoming Wait Time  During Call Wait Time  Calls  
0                  1.0                      4      1  
1                  1.0                      3      1  
2                  1.0                      3      1  
3                  1.0                      4      1  
4                  1.0                      4      1  
5                  1.0                      2      1  
6                  1.0                      3      1  
7                  1.0                      3      1  
8                  1.0                      4      1  
9                  1.0                      4      1  
10                 1.0                      4      1  
11                 1.0                      2      1  
12                 1.0          

In [2]:
south_df=pd.read_csv("SouthCallCenter.csv")
south_df.info()
south_df.head(20)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 314 entries, 0 to 313
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             314 non-null    int64  
 1   Branch                 314 non-null    object 
 2   Call Purpose           314 non-null    object 
 3   Time Block             314 non-null    object 
 4   Incoming or Outgoing   314 non-null    object 
 5   Queue                  314 non-null    object 
 6   Rep ID                 314 non-null    object 
 7   Sale                   314 non-null    object 
 8   Lost Call              314 non-null    int64  
 9   Abandoned              314 non-null    int64  
 10  Busy Minutes           314 non-null    int64  
 11  Not Ready Minutes      314 non-null    int64  
 12  Incoming Wait Time     126 non-null    float64
 13  During Call Wait Time  314 non-null    int64  
 14  Calls                  314 non-null    int64  
dtypes: flo

Unnamed: 0 Branch     Call Purpose Time Block Incoming or Outgoing Queue  \
0            0  South    Sales Support   10:00 AM             Incoming     D   
1            1  South    Sales Support   10:00 AM             Incoming     C   
2            2  South    Sales Support   10:00 AM             Incoming     C   
3            3  South    Sales Support   10:00 AM             Incoming     C   
4            4  South    Sales Support   12:00 PM             Incoming     C   
5            5  South    Sales Support    1:00 PM             Incoming     D   
6            6  South  Product Support   11:00 AM             Incoming     C   
7            7  South  Product Support   12:00 PM             Incoming     C   
8            8  South  Product Support    2:00 PM             Incoming     D   
9            9  South  Product Support    3:00 PM             Incoming     D   
10          10  South  Product Support    3:00 PM             Incoming     D   
11          11  South  Product Support    4:00 PM             Incoming     D   
12          12  South  Product Support    4:00 PM             Incoming     D   
13          13  South  Product Support    5:00 PM             Incoming     C   
14          14  South  Product Support    5:00 PM             Incoming     D   
15          15  South        Complaint   11:00 AM             Incoming     D   
16          16  South        Complaint   12:00 PM             Incoming     D   
17          17  South        Complaint    1:00 PM             Incoming     C   
18          18  South        Complaint    1:00 PM             Incoming     C   
19          19  South        Complaint    5:00 PM             Incoming     D   

    Rep ID Sale  Lost Call  Abandoned  Busy Minutes  Not Ready Minutes  \
0     Kate   NO          0          0             9                  1   
1     Eric   NO          0          0             8                  2   
2    Susan   NO          0          0            10                  2   
3    Alice   NO          0          0            12                  1   
4    Sandy   NO          0          0             8                  3   
5     Karl   NO          0          0             9                  1   
6    Sandy   NO          0          0             8                  1   
7    Sandy   NO          0          0             8                  1   
8    Randy   NO          0          0            11                  2   
9   George   NO          0          0             8                  3   
10   Helga   NO          0          0             9                  1   
11    Josh   NO          0          0            11                  3   
12  George   NO          0          0            10                  1   
13   Alice   NO          0          0            11                  3   
14   Randy   NO          0          0            10                  3   
15   Randy   NO          0          0            12                  3   
16    Josh   NO          0          0             9                  3   
17    Eric   NO          0          0            12                  3   
18   Sandy   NO          0          0            10                  3   
19    Karl   NO          0          0            12                  3   

    Incoming Wait Time  During Call Wait Time  Calls  
0                  1.0                      2      1  
1                  1.0                      4      1  
2                  1.0                      4      1  
3                  1.0                      3      1  
4                  1.0                      3      1  
5                  1.0                      4      1  
6                  1.0                      2      1  
7                  1.0                      3      1  
8                  1.0                      4      1  
9                  1.0                      4      1  
10                 1.0                      3      1  
11                 1.0                      3      1  
12                 1.0                      4      1 

<span style="background-color:orangered; color:orangered;">- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -</span>

## Exploratory Data Analysis

Time to do some EDA! In the process of learning more about the two datasets, answer the following questions. In the code blocks below enter the code used to solve the question.  At the end of the section record your answers by double clicking where asked.

#### EDA Question 1A:  How many reps are in the North branch?  

In [3]:
#Code Used:
north_df['Rep ID'].nunique()

9

#### EDA Question 1B:  How many reps are in the South branch?  

In [4]:
#Code Used:
south_df['Rep ID'].nunique()


11

#### EDA Question 2A:  What is the average busy minutes, not ready minutes, incoming wait time, and during call wait time for the North branch? 

In [5]:
#Code Used:
north_df.describe()

Unnamed: 0   Lost Call   Abandoned  Busy Minutes  Not Ready Minutes  \
count  245.000000  245.000000  245.000000    245.000000         245.000000   
mean   122.000000    0.020408    0.028571      9.987755           1.914286   
std     70.869599    0.141681    0.166940      1.403980           0.771416   
min      0.000000    0.000000    0.000000      8.000000           1.000000   
25%     61.000000    0.000000    0.000000      9.000000           1.000000   
50%    122.000000    0.000000    0.000000     10.000000           2.000000   
75%    183.000000    0.000000    0.000000     11.000000           3.000000   
max    244.000000    1.000000    1.000000     12.000000           3.000000   

       Incoming Wait Time  During Call Wait Time  Calls  
count           82.000000             245.000000  245.0  
mean             3.048780               2.967347    1.0  
std              1.456382               0.843827    0.0  
min              1.000000               2.000000    1.0  
25%              2.000000               2.000000    1.0  
50%              3.000000               3.000000    1.0  
75%              4.000000               4.000000    1.0  
max              5.000000               4.000000    1.0

#### EDA Question 2B:  What is the average busy minutes, not ready minutes, incoming wait time, and during call wait time for the South branch? 

In [6]:
#Code Used:
south_df.describe()

Unnamed: 0   Lost Call   Abandoned  Busy Minutes  Not Ready Minutes  \
count  314.000000  314.000000  314.000000     314.00000         314.000000   
mean   156.500000    0.035032    0.012739      10.05414           1.910828   
std     90.788215    0.184154    0.112324       1.40751           0.860016   
min      0.000000    0.000000    0.000000       8.00000           1.000000   
25%     78.250000    0.000000    0.000000       9.00000           1.000000   
50%    156.500000    0.000000    0.000000      10.00000           2.000000   
75%    234.750000    0.000000    0.000000      11.00000           3.000000   
max    313.000000    1.000000    1.000000      12.00000           3.000000   

       Incoming Wait Time  During Call Wait Time  Calls  
count          126.000000             314.000000  314.0  
mean             3.000000               3.076433    1.0  
std              1.308434               0.819424    0.0  
min              1.000000               2.000000    1.0  
25%              2.000000               2.000000    1.0  
50%              3.000000               3.000000    1.0  
75%              4.000000               4.000000    1.0  
max              5.000000               4.000000    1.0

#### EDA Question 3:  What is the number of calls taken for each time block(Both North and South combined)?

In [7]:
# Code Used:
# north_df.groupby('Time Block').Calls.sum()+south_df.groupby('Time Block').Calls.sum()
Time_Block_Calls = north_df.value_counts('Time Block')+south_df.value_counts('Time Block') 
Time_Block_Calls

Time Block
10:00 AM     99
11:00 AM     56
12:00 PM    120
1:00 PM      40
2:00 PM      65
3:00 PM      73
4:00 PM      53
5:00 PM      43
9:00 AM      10
dtype: int64

In [8]:
# checking total number of calls
north_df.Calls.sum()+south_df.Calls.sum()

559

##### Record your answers for the EDA section by double clicking here:
- 1a) Reps in the North Branch = 9
- 1b) Reps in the South Branch = 11


- 2a) North Branch, (round to the nearest hundredth):
> - Busy Minutes = 9.99 min
> - Not Ready Minutes = 1.91 min
> - Incoming Wait Time = 3.05 sec
> - During Call Wait Time = 2.97 min
- 2b) South Branch, (round to the nearest hundredth):
> - Busy Minutes = 10.05 min
> - Not Ready Minutes = 1.91 min
> - Incoming Wait Time = 3.00 sec
> - During Call Wait Time = 3.08 min
- 3) Total Calls taken for all branches per time block:
> - 9:00AM   = 10
> - 10:00AM  = 99
> - 11:00AM  = 56
> - 12:00PM  = 120
> - 1:00PM   = 40
> - 2:00PM   = 65
> - 3:00PM   = 73
> - 4:00PM   = 53
> - 5:00PM   = 43
> - Total Calls = 559


<span style="background-color:orangered; color:orangered;">- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -</span>

## Cleaning Your Data 
Now you need to clean up the datasets. When cleaning the datasets, you may find that there isn't dirty data to clean. That is okay! Some questions you need to answer about the data sets.
* Add additional code blocks as needed to show how you came to your conclusions. Add comments in your code blocks to help others understand your thinking. 

#### Cleaning Question 1:  Is there any missing data and if yes explain what you would do with the missing data and why?
##### Record your answer by double clicking here:
> Your Answer: The missing data appears in the "Incoming Wait Time" columnn for both the North and the South Call Centers. I have decided to leave these values as is as I cannot assume what the null value is suppose to be.

In [9]:
# Code Used
# Add Additional Cells as needed.
north_df.isnull().sum()

Unnamed: 0                 0
Branch                     0
Call Purpose               0
Time Block                 0
Incoming or Outgoing       0
Queue                      0
Rep ID                     0
Sale                       0
Lost Call                  0
Abandoned                  0
Busy Minutes               0
Not Ready Minutes          0
Incoming Wait Time       163
During Call Wait Time      0
Calls                      0
dtype: int64

In [10]:
# checking average
north_df['Incoming Wait Time'].nunique()

5

In [11]:
south_df.isnull().sum()

Unnamed: 0                 0
Branch                     0
Call Purpose               0
Time Block                 0
Incoming or Outgoing       0
Queue                      0
Rep ID                     0
Sale                       0
Lost Call                  0
Abandoned                  0
Busy Minutes               0
Not Ready Minutes          0
Incoming Wait Time       188
During Call Wait Time      0
Calls                      0
dtype: int64

In [12]:
# checking average
south_df['Incoming Wait Time'].nunique()

5

#### Cleaning Question 2:  In the North data set, there are two separate "YES" values for a sale. Why and how did you rectify the issue?
##### Record your answer by double clicking here:
> Your Answer: The "YES" values are listed as "YES" and "YES ". I will strip the extra spaces.

In [17]:
# finding the 2 different 'YES' values
north_df['Sale'].head(25)

0       NO
1       NO
2       NO
3       NO
4       NO
5       NO
6       NO
7       NO
8       NO
9       NO
10      NO
11      NO
12      NO
13      NO
14      NO
15      NO
16      NO
17      NO
18      NO
19    YES 
20      NO
21      NO
22      NO
23      NO
24      NO
Name: Sale, dtype: object

In [18]:
#Code Used
# Add Additional Cells as needed.

# checking to see number of each 'YES' value
north_df['Sale'].value_counts()


NO      234
YES      10
YES       1
Name: Sale, dtype: int64

In [19]:
# removing the extra space
north_df['Sale'] = north_df['Sale'].str.strip()

In [20]:
# checking to see if the is only 2 unique values: 'NO' and 'YES'
north_df['Sale'].nunique()

2

#### Cleaning Question 3:  Are there duplicates in the two data sets? If there are, how did you handle them?
##### Record your answer by double clicking here:
> Your Answer: There are no duplicates in either of the data sets. If there were, I would have removed them. 

In [21]:
#Code Used
# Add Additional Cells as needed.
north_df.duplicated().sum()

0

In [22]:
south_df.duplicated().sum()

0

#### Cleaning Question 4:  Is any of the data in the two data sets unnecessary? If yes, how did you handle it?
##### Record your answer by double clicking here:
> Your Answer: 
> - The "Unnamed" column is also unnecessary as it just reiterates the row number. I will drop both of these columns from each data set.

In [23]:
north_df['Unnamed: 0'].nunique()

245

In [24]:
south_df['Unnamed: 0'].nunique()

314

In [25]:
# checking to make sure the columns dropped
new_north_df = north_df.drop(columns=['Unnamed: 0'])
new_north_df

Branch     Call Purpose Time Block Incoming or Outgoing Queue  Rep ID  \
0    North        Complaint    9:00 AM             Incoming     B   Brent   
1    North    Sales Support   11:00 AM             Incoming     A     Cam   
2    North    Sales Support   11:00 AM             Incoming     B    Todd   
3    North    Sales Support    4:00 PM             Incoming     B   Brent   
4    North  Product Support   12:00 PM             Incoming     B   Brent   
..     ...              ...        ...                  ...   ...     ...   
240  North  Product Support    5:00 PM             Outgoing     B   Brent   
241  North  Product Support    5:00 PM             Outgoing     A     Joe   
242  North  Product Support    5:00 PM             Outgoing     A  Amanda   
243  North  Product Support    5:00 PM             Outgoing     A     Cam   
244  North  Product Support    5:00 PM             Outgoing     B  Xander   

    Sale  Lost Call  Abandoned  Busy Minutes  Not Ready Minutes  \
0     NO          0          1             9                  2   
1     NO          0          0            11                  1   
2     NO          0          0             9                  3   
3     NO          0          0            11                  2   
4     NO          0          0             8                  2   
..   ...        ...        ...           ...                ...   
240   NO          0          0            10                  2   
241   NO          0          0             8                  3   
242   NO          0          0            12                  1   
243   NO          0          0            11                  2   
244   NO          0          0             8                  1   

     Incoming Wait Time  During Call Wait Time  Calls  
0                   1.0                      4      1  
1                   1.0                      3      1  
2                   1.0                      3      1  
3                   1.0                      4      1  
4                   1.0                      4      1  
..                  ...                    ...    ...  
240                 NaN                      3      1  
241                 NaN                      2      1  
242                 NaN                      2      1  
243                 NaN                      2      1  
244                 NaN                      4      1  

[245 rows x 14 columns]

In [26]:
# checking to make sure the columns dropped
new_south_df = south_df.drop(columns=['Unnamed: 0'])
new_south_df

Branch     Call Purpose Time Block Incoming or Outgoing Queue  Rep ID  \
0    South    Sales Support   10:00 AM             Incoming     D    Kate   
1    South    Sales Support   10:00 AM             Incoming     C    Eric   
2    South    Sales Support   10:00 AM             Incoming     C   Susan   
3    South    Sales Support   10:00 AM             Incoming     C   Alice   
4    South    Sales Support   12:00 PM             Incoming     C   Sandy   
..     ...              ...        ...                  ...   ...     ...   
309  South  Product Support    5:00 PM             Outgoing     D   Helga   
310  South  Product Support    5:00 PM             Outgoing     C   Susan   
311  South  Product Support    5:00 PM             Outgoing     C   Sandy   
312  South  Product Support    5:00 PM             Outgoing     C    Eric   
313  South  Product Support    5:00 PM             Outgoing     D  George   

    Sale  Lost Call  Abandoned  Busy Minutes  Not Ready Minutes  \
0     NO          0          0             9                  1   
1     NO          0          0             8                  2   
2     NO          0          0            10                  2   
3     NO          0          0            12                  1   
4     NO          0          0             8                  3   
..   ...        ...        ...           ...                ...   
309   NO          0          0            10                  3   
310   NO          0          0            12                  3   
311   NO          0          0             9                  3   
312   NO          0          0            10                  1   
313   NO          0          0            10                  2   

     Incoming Wait Time  During Call Wait Time  Calls  
0                   1.0                      2      1  
1                   1.0                      4      1  
2                   1.0                      4      1  
3                   1.0                      3      1  
4                   1.0                      3      1  
..                  ...                    ...    ...  
309                 NaN                      4      1  
310                 NaN                      4      1  
311                 NaN                      4      1  
312                 NaN                      2      1  
313                 NaN                      3      1  

[314 rows x 14 columns]

<span style="background-color:orangered; color:orangered;">- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -</span>

## Data Manipulation
Before you begin answering any questions, combine the two datasets together to create a third dataframe called df. You can use this third dataframe to compare the two call centers to company-wide trends.
* Add additional code blocks as needed to show how you came to your conclusions. Add comments in your code blocks to help others understand your thinking. Record your answer by double clicking the below cell.

In [27]:
# Create dataframe for the entire company named df
both_branch_df=(pd.concat([new_north_df, new_south_df],
                  ignore_index = True))
both_branch_df

Branch     Call Purpose Time Block Incoming or Outgoing Queue  Rep ID  \
0    North        Complaint    9:00 AM             Incoming     B   Brent   
1    North    Sales Support   11:00 AM             Incoming     A     Cam   
2    North    Sales Support   11:00 AM             Incoming     B    Todd   
3    North    Sales Support    4:00 PM             Incoming     B   Brent   
4    North  Product Support   12:00 PM             Incoming     B   Brent   
..     ...              ...        ...                  ...   ...     ...   
554  South  Product Support    5:00 PM             Outgoing     D   Helga   
555  South  Product Support    5:00 PM             Outgoing     C   Susan   
556  South  Product Support    5:00 PM             Outgoing     C   Sandy   
557  South  Product Support    5:00 PM             Outgoing     C    Eric   
558  South  Product Support    5:00 PM             Outgoing     D  George   

    Sale  Lost Call  Abandoned  Busy Minutes  Not Ready Minutes  \
0     NO          0          1             9                  2   
1     NO          0          0            11                  1   
2     NO          0          0             9                  3   
3     NO          0          0            11                  2   
4     NO          0          0             8                  2   
..   ...        ...        ...           ...                ...   
554   NO          0          0            10                  3   
555   NO          0          0            12                  3   
556   NO          0          0             9                  3   
557   NO          0          0            10                  1   
558   NO          0          0            10                  2   

     Incoming Wait Time  During Call Wait Time  Calls  
0                   1.0                      4      1  
1                   1.0                      3      1  
2                   1.0                      3      1  
3                   1.0                      4      1  
4                   1.0                      4      1  
..                  ...                    ...    ...  
554                 NaN                      4      1  
555                 NaN                      4      1  
556                 NaN                      4      1  
557                 NaN                      2      1  
558                 NaN                      3      1  

[559 rows x 14 columns]

#### Manipulation Question 1:  Group by Rep ID and sum the resulting structure. Sort by calls to determine which rep in each branch has the highest number of calls.

##### Record your answer by double clicking here:
Rep with the hightest number of calls and their total calls:
- North Branch = Brent - 37 calls
- South Branch = Eric - 35 calls


In [28]:
# Code Used
# Add Additional Cells as needed.
new_north_df.groupby("Rep ID").sum().sort_values(by='Calls', ascending=False)

C:\Users\kelle\AppData\Local\Temp\ipykernel_11184\80170868.py:3: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  new_north_df.groupby("Rep ID").sum().sort_values(by='Calls', ascending=False)


Lost Call  Abandoned  Busy Minutes  Not Ready Minutes  \
Rep ID                                                          
Brent           0          1           362                 71   
Lilly           0          2           307                 59   
Duke            2          0           292                 56   
Todd            1          0           296                 58   
Joe             0          0           256                 49   
Cam             1          0           233                 46   
Xander          1          0           244                 44   
Amanda          0          1           229                 43   
Andy            0          3           228                 43   

        Incoming Wait Time  During Call Wait Time  Calls  
Rep ID                                                    
Brent                 17.0                    117     37  
Lilly                 37.0                     88     30  
Duke                  39.0                     85     29  
Todd                  21.0                     97     29  
Joe                   36.0                     81     26  
Cam                   32.0                     66     24  
Xander                20.0                     68     24  
Amanda                12.0                     60     23  
Andy                  36.0                     65     23

In [29]:
new_south_df.groupby("Rep ID").sum().sort_values(by='Calls', ascending=False)

C:\Users\kelle\AppData\Local\Temp\ipykernel_11184\2800984857.py:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  new_south_df.groupby("Rep ID").sum().sort_values(by='Calls', ascending=False)


Lost Call  Abandoned  Busy Minutes  Not Ready Minutes  \
Rep ID                                                          
Eric            0          0           361                 54   
Randy           3          1           320                 66   
Sandy           1          0           325                 60   
Helga           0          2           292                 58   
George          0          0           298                 53   
Kate            0          0           291                 61   
Karl            1          0           287                 61   
Josh            3          0           260                 50   
Susan           2          0           267                 53   
Sharon          1          1           257                 38   
Alice           0          0           199                 46   

        Incoming Wait Time  During Call Wait Time  Calls  
Rep ID                                                    
Eric                  47.0                    112     35  
Randy                 47.0                     98     33  
Sandy                 34.0                    100     32  
Helga                 36.0                     95     31  
George                41.0                     87     29  
Kate                  37.0                     83     29  
Karl                  26.0                     85     28  
Josh                  31.0                     83     26  
Susan                 35.0                     88     26  
Sharon                29.0                     77     25  
Alice                 15.0                     58     20

#### Manipulation Question 2:  The average call abandonment rate is a KPI when it comes to determining call center efficiency. As you may recall, abandoned calls are calls where the customer hangs up due to long call times. What is the average call abandonment rate for each branch and the whole company? Do any of these fall out of the optimal range of 2-5%?
##### Record your answer by double clicking here:
Average Call Abandonment Rates (round to the nearest hundredth):
- North Branch = .03
- South Branch = .01
- Company Wide = .02
- Do any of these fall out of the optimal range of 2-5%? NO

In [30]:
# Code Used
# Add Additional Cells as needed
new_north_df["Abandoned"].mean()

0.02857142857142857

In [31]:
new_south_df["Abandoned"].mean()

0.012738853503184714

In [32]:
both_branch_df["Abandoned"].mean()

0.01967799642218247

#### Manipulation Question 3:  Service level is another KPI when it comes to measuring call center efficiency. Service level is the percentage of calls answered within a specific number of seconds. In the case of your employer, their ideal time frame is 2 seconds. What is the percentage of calls answered within 2 seconds for each branch and the entire company?
##### Record your answer by double clicking here:
Percentage of calls answered within 2 seconds, include # of calls:
- North Branch = 40.24% / 33 calls under 2 seconds
- South Branch = 38.10% / 48 calls under 2 seconds
- Company Wide = 38.94% / 81 calls under 2 seconds

In [36]:
# Code Used
# Add Additional Cells as needed.
incoming_wait_times_north = new_north_df['Incoming Wait Time']<=2
incoming_wait_times_north.value_counts()


False    212
True      33
Name: Incoming Wait Time, dtype: int64

In [37]:
incoming_wait_times_north.sum()/82*100

40.243902439024396

In [38]:
incoming_wait_times_south = new_south_df['Incoming Wait Time']<=2
incoming_wait_times_south.value_counts()

False    266
True      48
Name: Incoming Wait Time, dtype: int64

In [39]:
incoming_wait_times_south.sum()/126*100

38.095238095238095

In [40]:
both_incoming_wait_time = both_branch_df['Incoming Wait Time']<=2
both_incoming_wait_time.value_counts()

False    478
True      81
Name: Incoming Wait Time, dtype: int64

In [41]:
both_incoming_wait_time.sum()/208*100

38.94230769230769

#### Manipulation Question 4: For each branch and the entire company, what is the average speed of answer?
##### Record your answer by double clicking here:
Average speed of answer (rounded to nearest hundredth):
- North Branch in seconds = 1.02
- South Branch in seconds = 3.00
- Company Wide in seconds = 1.69


In [42]:
# Code Used
new_north_df['Incoming Wait Time'].mean()


3.048780487804878

In [43]:
new_south_df['Incoming Wait Time'].mean()

3.0

In [44]:
both_branch_df['Incoming Wait Time'].mean()

3.019230769230769

<span style="background-color:orangered; color:orangered;">- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -</span>

## Visualization

Create a visualization for each of the following questions. Some of the code to handle aggregating and storing data may be written for you. For each visualization, you choose the chart style that you feel suits the situation best. Make note of the chart style you chose and why. 

*NOTE Some questions you may decide to use more than one chart and or chart style.

#### Visualization 1:  What is the average abandonment rate per queue?
##### Record your answer by double clicking here:

> Chart style you chose and why: 

In [47]:
# Create visualization 1 here
# Import any additional libraries needed with alias
import seaborn as sns


# The dictionary abandonment_rates has the data you need.
abandonment_rates = {}
queues = ["A", "B", "C", "D"]
queue_dict = both_branch_df.groupby("Queue").agg("sum")
for i in range(2):
    abandonment_rates[queues[i]] = queue_dict["Abandoned"][i] / queue_dict["Calls"][i] 

#Your Code Here:



C:\Users\kelle\AppData\Local\Temp\ipykernel_11184\15651337.py:9: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  queue_dict = both_branch_df.groupby("Queue").agg("sum")


#### Visualization 2: What is the service level percentage and average speed of answer for each rep in the North Branch?
##### Record your answer by double clicking here:

> Chart style you chose and why: 

In [46]:
# Create visualization 2 here
# north_plt contains the data you need for the average speed of answer of each rep

north_plt = new_north_df.groupby("Rep ID")["Incoming Wait Time"].mean().to_frame().reset_index()

# Finding each Rep's Personal Service Level Percentage.  Basically, Calls within 2 secs / total calls

# Table 1: Total Incoming calls less than 2 seconds grouped by Rep
quick_calls = new_north_df[new_north_df["Incoming Wait Time"] <= 2.0]
quick_reps = quick_calls[["Rep ID", "Calls"]]
quick_stats = quick_reps.groupby(["Rep ID"]).sum()  # Final Table


# Table 2: Total Incoming Calls Only grouped by Rep
total_calls_in = new_north_df[new_north_df["Incoming or Outgoing"] == "Incoming"]
rep_calls = total_calls_in[["Rep ID", "Calls"]]     
total_stats = rep_calls.groupby(["Rep ID"]).sum() # Final Table  

#  Table 3: Service Level Percentage created via merge
service_level = pd.merge(quick_stats, total_stats, on="Rep ID")

# Create Percentage Column in Table 3
service_level["Service Level %"] = service_level["Calls_x"]/service_level["Calls_y"] * 100

#Your Code Here:


#### Visualization 3: For each type of call purpose, how many calls are outgoing vs. incoming?
##### Record your answer by double clicking here:

> Chart style you chose and why: 

In [45]:
# Create visualization 3 here:
# The three dictionaries, complaints, sales_support, and product_support, have the information you need

purpose_group = both_branch_df.groupby("Call Purpose")
call_purpose = ["Complaint", "Product Support", "Sales Support"]
purpose_counts = purpose_group["Incoming or Outgoing"].value_counts()
print(purpose_counts)

complaints = purpose_counts["Complaint"].to_dict()
sales_support = purpose_counts["Sales Support"].to_dict()
product_support = purpose_counts["Product Support"].to_dict()

#Your Code Here:


Call Purpose     Incoming or Outgoing
Complaint        Incoming                 51
Product Support  Outgoing                228
                 Incoming                 82
Sales Support    Outgoing                123
                 Incoming                 75
Name: Incoming or Outgoing, dtype: int64


<span style="background-color:orangered; color:orangered;">- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -</span>

## Summarize Your Work

With what you know now about the two call centers and the entire company, answer the following questions. Note that while this is subjective, you should include relevant data to back up your opinion.


#### Summary Question 1:  Using KPIs such as average abandonment rate, service level and average speed of answer, in your opinion, which one of the two branches is operating more efficiently? Why?

##### Record your answer by double clicking here:
> Your Answer: 



#### Summary Question 2: Based on the number of reps in each branch and how quickly the reps are working, in your opinion, which branch would benefit from the extra help?

##### Record your answer by double clicking here:
> Your Answer: 


#### Summary Question 3: Now that you have explored the datasets, is there any data or information that you wish you had in this analysis?

##### Record your answer by double clicking here:
> Your Answer: 


<span style="background-color:orangered; color:orangered;">- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -</span>

## Bonus Mission
Create a visualization that answers this question: For each call purpose, how many calls (incoming and outgoing) take place in each time block?
##### Record your answer by double clicking here:

> Chart style you chose and why: 

In [ ]:
# Create your Bonus Mission visualization here!
call_times = df[["Time Block", "Call Purpose", "Incoming or Outgoing", "Calls"]]

# Use groupby to plot based on time blocks:

# Use groupby and get_group to select which call purpose to plot: